### ISSUE TITLE
The Makefile has a bug: POSIX shell standard is awful, part 0x666

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#1085]
TL;DR:

This is a story about the horrors of the POSIX shell standard and how those problems can affect a far saner shell. I noticed this while working on a change that failed a unit test. Which caused `make test` to run `go test` twice when it should have done so only once.

Long version:

@xiaq recently modified the *Makefile* in response to my reporting that `make test` fails on 32-bit platforms since `go test -race` does not work on those platforms. The problem is that change uses a POSIX shell construct that devolves to this if you introduce a change that breaks `make test` on a platform that supports `go test -race`:

```
true && false || echo WTF
```

The `true` represents the command that determines if `go test -race` works.

The `false` represents the `go test -race` that fails due to an actual bug in the elvish code.

The `echo WTF` represents the `go test` without a `-race` option. It is not intended to run in that case **but it does**. That is because the author, and myself, thought the above was equivalent to this:

```
if true; then
    false
else
    echo WTF
fi
```

That interpretation is wrong. It causes the tests to be run twice on a platform which supports `go test -race` and at least one test fails. See the [POSIX shell standard](https://pubs.opengroup.org/onlinepubs/007908799/xcu/chap2.htm) section regarding "Lists" of commands. There isn't a single example of what happens when the second "list" of commands fails. Which is one reason the incorrect expectation of how `a && b || c` behaves is so prevalent.

I actually reviewed the problematic change and did not notice the bug. I did, however, point out an insignificant performance optimization. This is a good example of one problem, out of hundreds, with the POSIX shell standard. Which is why alternative shells, like Elvish, exist.

**Comment by xiaq** [comment]
Well, I guess the lesson is "don't use a && b || c as a shorthand for if/else". As tricky as it is, this is a problem of misuse.

**Comment by xiaq** [comment]
https://www.shellcheck.net/ catches this issue, it has a warning when you use `a && b || c`.

**Comment by occivink** [comment]
I'm honestly not sure why you're saying that this is a case of posix horror. Most programming languages will evaluate this particular condition in the same way, I don't know why you expect the shell to be different. 
The lesson is more that you shouldn't try to cram logic into tricky one-liners.

**Comment by krader1961** [comment]
@occivink, The point is that I've seen the `&&` and `||` command list operators create bugs of this nature so many times over the past three decades I've long since lost count. At work I never allow their use in a script. If you do think saving of a few characters is worthwhile they should be limited to only a single `&&` or `||`, and never use them together in the same command list.

Yes, those operators are not the worst thing about the POSIX shell standard. For that honor I would point to the `IFS` var and the rules for splitting text into tokens after variable expansion if the expansion is not enclosed in double-quotes. :smile: The point is that these operators behave like logical "and" and "or" in other languages but that isn't what (nearly) everyone intuitively expects when they see `cmd-a && cmd-b || cmd-c`. All of which is to say that I hope Elvish never implements those command list operators.
